### PR TITLE
Fix a false negative of the [<TailCall>] analysis in combination with async (issue #17237)

### LINKED ISSUE #17237: Incorrect TailCall warning on async recursive functions
When we use the TailCall attribute on an async recursive function, that actually has an async call, we get a warning that it's not being used in a tail recursive way, but it should be

Examples below:
```
[<TailCall>] //Has Warning
let rec f'1 (g: bool Async) = async {
    match! g with
    | false -> ()
    | true -> return! f'1 g
    }


[<TailCall>] //Has Warning
let rec f'2 (g: bool Async) = async {
    let! x = g
    match x with
    | false -> ()
    | true -> return! f'2 g
    }

[<TailCall>] //No Warning
let rec f'3 (g: bool) = async {
    match g with
    | false -> ()
    | true -> return! f'3 g
    }
```

**Repro steps**

1. Create an f# project
2. Create a recursive function that has an async call and use the TailCall attribute.
4. Build, and notice the warning

Sample sln: 
[TailCallTest.zip](https://github.com/dotnet/fsharp/files/15419262/TailCallTest.zip)

**Expected behavior**
No warnings

**Actual behavior**
Warnings display in output
>warning FS3569: The member or function 'f'1' has the 'TailCallAttribute' attribute, but is not being used in a tail recursive way.
>warning FS3569: The member or function 'f'2' has the 'TailCallAttribute' attribute, but is not being used in a tail recursive way.

**Known workarounds**
Remove the TailCall attribute, but I am unsure if this is actually running as a TailCall

**Related information**

Provide any related information (optional):
Occurs in .net8.0, .net6.0, .net5.0
Visual Studio 17.9.7 and 17.10.0

### PR DESCRIPTION
**Pull request by dawedawe**
## Description

Fixes #17237

## Checklist

- [x] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [x] Release notes entry updated:

### CODE REVIEW

File: src/Compiler/Checking/TailCallChecks.fs [hunk 1]
```diff
@@ -756,7 +756,25 @@ let CheckModuleBinding cenv (isRec: bool) (TBind _ as bind) =
                 | Expr.Lambda(bodyExpr = bodyExpr) -> checkTailCall insideSubBindingOrTry bodyExpr
                 | Expr.DebugPoint(_debugPointAtLeafExpr, expr) -> checkTailCall insideSubBindingOrTry expr
                 | Expr.Let(binding = binding; bodyExpr = bodyExpr) ->
-                    checkTailCall true binding.Expr
+                    let isAsyncContinuation =
+                        match bodyExpr with
+                        | Expr.App(funcExpr = Expr.Val(valRef = valRef)) ->
```
Comment (dawedawe) [inline src/Compiler/Checking/TailCallChecks.fs:761]: Not super sure about this. A simpler test might be:
```
when
                            valRef.DisplayNameCore = "MakeAsync"
                            && valRef.HasDeclaringEntity
                            && valRef.DeclaringEntity.LogicalName = "AsyncPrimitives"
```

But that might be too much cowboy-style :cowboy_hat_face:

Comment (T-Gro) [reply src/Compiler/Checking/TailCallChecks.fs:761]: I don't think we should hardcode implementation specifics (from FSharp.Core) into compiler checking.

If there is no other way to detect this as of now, I would have another proposal:
Make it a reusable component, e.g. an attribute place on the decleration side (in this case, within AsyncPrimitives implementation) to make it less hardcoded.

Comment (vzarytovskii) [inline src/Compiler/Checking/TailCallChecks.fs:761]: Are we 100% sure resulting code will be tail recursive?
```fsharp
namespace N
module M =
    [<TailCall>]
    let rec f (g: bool Async) =
        async {
            match! g with
            | false -> ()
            | true -> return! f g
        }

    [<EntryPoint>]
    let main _ =
        f ( async { return true } ) |> Async.RunSynchronously
        0
```

Will result in chained calls of various generated closure classes:
```csharp
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using Microsoft.FSharp.Control;
using Microsoft.FSharp.Core;

namespace N;

[CompilationMapping(SourceConstructFlags.Module)]
public static class M
{
    [Serializable]
    internal sealed class f@6-1 : FSharpFunc<bool, FSharpAsync<Unit>>
    {
        public FSharpAsync<bool> g;

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal f@6-1(FSharpAsync<bool> g)
        {
            this.g = g;
        }

        public override FSharpAsync<Unit> Invoke(bool _arg1)
        {
            if (!_arg1)
            {
                return ExtraTopLevelOperators.DefaultAsyncBuilder.Zero();
            }
            return f(g);
        }
    }

    [Serializable]
    internal sealed class f@6-2 : FSharpFunc<AsyncActivation<Unit>, AsyncReturn>
    {
        public FSharpAsync<bool> g;

        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
        [CompilerGenerated]
        [DebuggerNonUserCode]
        public FSharpFunc<bool, FSharpAsync<Unit>> binder;

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal f@6-2(FSharpAsync<bool> g, FSharpFunc<bool, FSharpAsync<Unit>> binder)
        {
            this.g = g;
            this.binder = binder;
        }

        public override AsyncReturn Invoke(AsyncActivation<Unit> ctxt)
        {
            return AsyncPrimitives.Bind(ctxt, g, binder);
        }
    }

    [Serializable]
    internal sealed class f@6 : FSharpFunc<Unit, FSharpAsync<Unit>>
    {
        public FSharpAsync<bool> g;

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal f@6(FSharpAsync<bool> g)
        {
            this.g = g;
        }

        public override FSharpAsync<Unit> Invoke(Unit unitVar)
        {
            FSharpFunc<bool, FSharpAsync<Unit>> binder = new f@6-1(g);
            return AsyncPrimitives.MakeAsync(new f@6-2(g, binder));
        }
    }

    [Serializable]
    internal sealed class main@13-1 : FSharpFunc<AsyncActivation<bool>, AsyncReturn>
    {
        internal static readonly main@13-1 @_instance = new main@13-1();

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal main@13-1()
        {
        }

        public override AsyncReturn Invoke(AsyncActivation<bool> ctxt)
        {
            return AsyncActivation<bool>.Success(ctxt, result: true);
        }
    }

    [Serializable]
    internal sealed class main@13 : FSharpFunc<Unit, FSharpAsync<bool>>
    {
        internal static readonly main@13 @_instance = new main@13();

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal main@13()
        {
        }

        public override FSharpAsync<bool> Invoke(Unit unitVar)
        {
            return AsyncPrimitives.MakeAsync(main@13-1.@_instance);
        }
    }

    [TailCall]
    public static FSharpAsync<Unit> f(FSharpAsync<bool> g)
    {
        return ExtraTopLevelOperators.DefaultAsyncBuilder.Delay(new f@6(g));
    }

    [EntryPoint]
    public static int main(string[] _arg1)
    {
        FSharpAsync<Unit> computation = f(ExtraTopLevelOperators.DefaultAsyncBuilder.Delay(main@13.@_instance));
        FSharpAsync.RunSynchronously(computation, null, null);
        return 0;
    }
}
```

Comment (vzarytovskii) [reply src/Compiler/Checking/TailCallChecks.fs:761]: I suppose it is since we call f -> f@6 -> f@6-2 -> 6@6-1 -> f or something.

Comment (dawedawe) [reply src/Compiler/Checking/TailCallChecks.fs:761]: Yes, from what I see, it's tail recursive.

File: src/Compiler/Checking/TailCallChecks.fs [hunk 2]
```diff
@@ -756,7 +756,18 @@ let CheckModuleBinding cenv (isRec: bool) (TBind _ as bind) =
                 | Expr.Lambda(bodyExpr = bodyExpr) -> checkTailCall insideSubBindingOrTry bodyExpr
                 | Expr.DebugPoint(_debugPointAtLeafExpr, expr) -> checkTailCall insideSubBindingOrTry expr
                 | Expr.Let(binding = binding; bodyExpr = bodyExpr) ->
-                    checkTailCall true binding.Expr
+                    // detect continuation shapes like MakeAsyns
```
Comment (abonie) [inline src/Compiler/Checking/TailCallChecks.fs:759]:
```suggestion
                    // detect continuation shapes like MakeAsync
```

**Review by abonie — APPROVED** [review]

**Review by T-Gro — APPROVED** [review]

**Pull request merged**